mava/networks/mat_network.py

```diff
@@ -0,0 +1,317 @@
+    continuous_autoregressive_act,
+    continuous_parallel_act,
+    discrete_autoregressive_act,
+    discrete_parallel_act,
+)
+
+
+class SwiGLU(nn.Module):
+    ffn_dim: int
+    embed_dim: int
+
+    def setup(self) -> None:
+        self.W_1 = self.param("W_1", nn.initializers.zeros, (self.embed_dim, self.ffn_dim))
+        self.W_G = self.param("W_G", nn.initializers.zeros, (self.embed_dim, self.ffn_dim))
+        self.W_2 = self.param("W_2", nn.initializers.zeros, (self.ffn_dim, self.embed_dim))
+
+    def __call__(self, x: chex.Array) -> chex.Array:
+        return (jax.nn.swish(x @ self.W_G) * (x @ self.W_1)) @ self.W_2
+
+
+class EncodeBlock(nn.Module):
+    n_embd: int
+    n_head: int
+    n_agent: int
+    masked: bool = False
+    use_swiglu: bool = False
+    use_rmsnorm: bool = False
+
+    def setup(self) -> None:
+        ln = nn.RMSNorm if self.use_rmsnorm else nn.LayerNorm
+        self.ln1 = ln()
+        self.ln2 = ln()
+
+        self.attn = SelfAttention(self.n_embd, self.n_head, self.n_agent, self.masked)
+
+        if self.use_swiglu:
+            self.mlp = SwiGLU(self.n_embd, self.n_embd)
+        else:
+            self.mlp = nn.Sequential(
+                [
+                    nn.Dense(self.n_embd, kernel_init=orthogonal(jnp.sqrt(2))),
+                    nn.gelu,
+                    nn.Dense(self.n_embd, kernel_init=orthogonal(0.01)),
+                ],
+            )
+
+    def __call__(self, x: chex.Array) -> chex.Array:
+        x = self.ln1(x + self.attn(x, x, x))
+        x = self.ln2(x + self.mlp(x))
+        return x
+
+
+class Encoder(nn.Module):
+    obs_dim: int
+    action_dim: int
+    n_block: int
+    n_embd: int
+    n_head: int
+    n_agent: int
+    use_swiglu: bool = False
+    use_rmsnorm: bool = False
+
+    def setup(self) -> None:
+        ln = nn.RMSNorm if self.use_rmsnorm else nn.LayerNorm
+
+        self.obs_encoder = nn.Sequential(
+            [ln(), nn.Dense(self.n_embd, kernel_init=orthogonal(jnp.sqrt(2))), nn.gelu],
+        )
+        self.ln = ln()
+        self.blocks = nn.Sequential(
+            [
+                EncodeBlock(
+                    self.n_embd,
+                    self.n_head,
+                    self.n_agent,
+                    use_swiglu=self.use_swiglu,
+                    use_rmsnorm=self.use_swiglu,
+                )
+                for _ in range(self.n_block)
+            ]
+        )
+        self.head = nn.Sequential(
+            [
+                nn.Dense(self.n_embd, kernel_init=orthogonal(jnp.sqrt(2))),
+                nn.gelu,
+                ln(),
+                nn.Dense(1, kernel_init=orthogonal(0.01)),
+            ],
+        )
+
+    def __call__(self, obs: chex.Array) -> Tuple[chex.Array, chex.Array]:
+        obs_embeddings = self.obs_encoder(obs)
+        x = obs_embeddings
+
+        rep = self.blocks(self.ln(x))
+        v_loc = self.head(rep)
+
+        return jnp.squeeze(v_loc, axis=-1), rep
```

Comment:
Never understood why this was call v_loc, I think value is more explicit?
```suggestion
        value = self.head(rep)

        return jnp.squeeze(value, axis=-1), rep
```

Reply:
Just got it from the original implementation :sweat_smile:
Renamed.